### ISSUE TITLE
1.2.9 deflateParams API change

### ISSUE THREAD

**Issue by pmqs** [madler/zlib#190]
My Perl interface to zlib exposes the deflateParams function. This is currently broken with 1.2.9.

Looking at the difference in the documentaion for deflateParams between the 1.2.8 and 1.2.9 it looks like a backward incompatible API change.

This is what 1.2.8 says about deflateParams

```
     Before the call of deflateParams, the stream state must be set as for
   a call of deflate(), since the currently available input may have to be
   compressed and flushed.  In particular, strm->avail_out must be non-zero.
```

Note the final sentence. I used that in my code so I could get by with setting strm->avail_out to 1. That meant that the internal state in my mapping code only had to remember one extra compressed byte to output with the next call to deflate.

zlib 1.2.9 seems to have changed that and I'm seeing deflateParams fail with Z_BUF_ERROR. That ties in with the new documentation for deflateParams as it now expects to be able to flush all internal data to the output buffer.

I can code around the API change, but wonder if this change will break anything else?

**Comment by madler** [comment]
Yes, this was a necessary change since the API itself had an insurmountable bug.

**Comment by abucodonosor** [comment]
Since you broke the API you should bump the lib version +1.. No matter the old API was buggy or not .. API changes soname++ ..

**Comment by pmqs** [comment]
Is this the only API change in 1.2.9 that has a backward compatibility implications?

I think a document that explicitly calls out the changes to the API that have the potential to break existing code would help.

**Comment by madler** [comment]
I have remedied one aspect of this change, so that now in 1.2.11 a `deflateParms()` before any data has been consumed by `deflate()` will immediately change the parameters. E.g. right after `deflateInit()`.

**Comment by praiskup** [comment]
Sorry if I miss something, but I hope I finally got to the point of this report.

Is this bugreport related to f1ebdd6a9c495a5db9a22aa80dd7d54ae7db42e9 ?

**Comment by madler** [comment]
No, this issue has no relation to that commit.

**Comment by praiskup** [comment]
Yeah, sorry.  We had a downstream patch  http://pkgs.fedoraproject.org/cgit/rpms/zlib.git/tree/zlib-1.2.7-z-block-flush.patch related to deflateParams() which was related to f1ebdd6, but this issue was later fixed by c661c374e8df4336 upstream.  I'm going to drop the downstream patch then.
